kernel/src/table_features/mod.rs

```diff
@@ -220,6 +598,206 @@ pub(crate) static SUPPORTED_WRITER_FEATURES: LazyLock<Vec<TableFeature>> = LazyL
+        let mut features = HashSet::new();
+
+        // Add all reader features if present
+        if let Some(reader_features) = self.protocol.reader_features() {
+            for feature in reader_features {
+                if self.is_supported(feature) {
+                    features.insert(feature.clone());
+                }
+            }
+        }
+
+        // Add all writer features if present
+        if let Some(writer_features) = self.protocol.writer_features() {
+            for feature in writer_features {
+                if self.is_supported(feature) {
+                    features.insert(feature.clone());
+                }
+            }
+        }
+
+        features
+    }
+
+    /// Helper method to check if a feature appears in the protocol's feature lists.
+    /// This is used for table features protocol (min_reader >= 3 && min_writer >= 7).
+    pub(crate) fn is_in_feature_lists(&self, feature: &TableFeature) -> bool {
+        let feature_type = feature.feature_type();
+
+        match feature_type {
+            FeatureType::Writer => {
+                // Writer-only features must appear in writer_features
+                self.protocol
+                    .writer_features()
+                    .map(|features| features.contains(feature))
+                    .unwrap_or(false)
+            }
+            FeatureType::ReaderWriter => {
+                // ReaderWriter features must appear in both lists
+                let in_reader = self
+                    .protocol
+                    .reader_features()
+                    .map(|features| features.contains(feature))
+                    .unwrap_or(false);
+                let in_writer = self
+                    .protocol
+                    .writer_features()
+                    .map(|features| features.contains(feature))
+                    .unwrap_or(false);
+
+                in_reader && in_writer
+            }
+            FeatureType::Unknown => false,
+        }
+    }
+
+    /// Validates that the Rust kernel has read support for all features required by the table.
+    ///
+    /// Returns an error if any ReaderWriter feature lacks read support in the kernel.
+    /// Writer-only features are not checked since they don't affect reading.
+    pub(crate) fn validate_read_support(&self) -> DeltaResult<()> {
+        let supported_features = self.get_supported_features();
+
+        for feature in supported_features {
+            if let Some(info) = feature.info() {
+                // Only check ReaderWriter features for read support
+                if matches!(info.feature_type, FeatureType::ReaderWriter) && !info.has_read_support
+                {
+                    return Err(Error::unsupported(format!(
+                        "Table feature '{}' is not supported for reading",
+                        feature
+                    )));
+                }
+            }
```

Comment:
We may need a better story for this, because "is it supported" is usually more complicated than a yes/no.

Example: Scan's supported reading Column mapping tables, but CDF didn't for the longest time. 


Or how kernel will be able to write to CDF-enabled tables, but only when AppendOnly=true.